packages/itwin/grouping-mapping-widget/src/test/QueryBuilder.test.ts

```diff
@@ -0,0 +1,186 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+
+import { Query, QueryBuilder, QueryClass } from "../widget/components/QueryBuilder";
+import { assert } from "chai"
+
+describe("QueryBuilder", () => {
+  describe("buildQueryString()", () => {
+
+    beforeEach(() => {
+    });
+
+    afterEach(() => {
+    });
+
+    const testBuildQueryString = (query: Query|undefined, expectedResult: string) => {
+      let sut = new QueryBuilder(undefined);
+      sut.query = query
+      let result = sut.buildQueryString();
+
+      assert.strictEqual(result, expectedResult)
+    }
+
+    it("should return empty string, when query is undefined", () => testBuildQueryString(undefined, ""))
+    it("should return empty string, when query unions are empty", () => testBuildQueryString({unions: []}, ""))
+    it("should return empty string, when query classes are empty", () => testBuildQueryString({unions: [{classes: []}]}, ""))
+
+    it("should return query string with ROUND, when property is float", () => {
+      let baseClass: QueryClass = {
+        className: "BaseClassName",
+        isAspect: false,
+        isRelational: true,
+        properties: []
+      };
+
+      let joinClass: QueryClass = {
+        className: "JoinClassName",
+        isAspect: false,
+        isRelational: true,
+        properties: [
+          {
+            name: "propName1",
+            needsQuote: false,
+            isCategory: false,
+            value: 3.14159
+          }
+        ]
+      };
+
+      let query: Query = {
+        unions: [
+          {classes: [baseClass, joinClass]}
+        ]
+      }
+
+      testBuildQueryString(query, "SELECT BaseClassName.ECInstanceId FROM BaseClassName JOIN JoinClassName ON ROUND(JoinClassName.propName1, 4)=3.1416")
+    });
+
+    it("should return query string with propery value in quotes, when property needs quotes", () => {
+      let baseClass: QueryClass = {
+        className: "BaseClassName",
+        isAspect: false,
+        isRelational: true,
+        properties: []
+      };
+
+      let joinClass: QueryClass = {
+        className: "JoinClassName",
+        isAspect: false,
+        isRelational: true,
+        properties: [
+          {
+            name: "propName1",
+            needsQuote: true,
+            isCategory: false,
+            value: "someName"
+          }
+        ]
+      };
+
+      let query: Query = {
+        unions: [
+          {classes: [baseClass, joinClass]}
+        ]
+      }
+
+      testBuildQueryString(query, "SELECT BaseClassName.ECInstanceId FROM BaseClassName JOIN JoinClassName ON JoinClassName.propName1='someName'")
```

Comment:
These 'ON' clauses make no sense, because there is no reference to the `BaseClassName`. How do the rows from the `join`ed class link to the `from` class? There is no mapping.

FYI: Do not fix it in this PR, separate into another one. This one is solving a specific problem and let's keep it that way. Just wanted to let you know that this specific test will need to be changed.

Reply:
Will look into it in other PRs